frame/merkle-mountain-range/src/mmr/storage.rs

```diff
@@ -112,11 +179,55 @@ where
 		let mut leaf_index = leaves;
 		let mut node_index = size;
 
+		// Use parent hash of block adding new nodes (this block) as extra identifier
+		// in offchain DB to avoid DB collisions and overwrites in case of forks.
+		let parent_hash = <frame_system::Pallet<T>>::parent_hash();
+		let block_number = <frame_system::Pallet<T>>::block_number();
 		for elem in elems {
+			if helper::pos_height_in_tree(node_index) == 0 {
+				info!(
+					target: "runtime::mmr",
+					"🥩: appending leaf {}, pos {}, block-num {:?}",
+					leaf_index, node_index, block_number
+				);
+			} else {
+				info!(
+					target: "runtime::mmr",
+					"🥩: appending node pos {}, block-num {:?}",
+					node_index, block_number
+				);
+			}
+			let key = Pallet::<T, I>::offchain_key(parent_hash, node_index);
+			info!(
+				target: "runtime::mmr",
+				"🥩: offchain set: pos {} parent_hash {:?} key {:?}",
+				parent_hash, node_index, key
+			);
 			// Indexing API is used to store the full node content (both leaf and inner).
-			elem.using_encoded(|elem| {
-				offchain_index::set(&Pallet::<T, I>::offchain_key(node_index), elem)
-			});
+			elem.using_encoded(|elem| offchain_index::set(&key, elem));
+
+			// Now we need to "canonicalize" any old leaves.
+			let window_size = <T as frame_system::Config>::BlockHashCount::get()
+				.saturating_sub(One::one())
+				.unique_saturated_into();
+			if leaf_index >= window_size && helper::pos_height_in_tree(node_index) == 0 {
+				let leaves = leaf_index + 1;
+				// move the rolling window for `block_num->hash` mappings available in the runtime:
+				// on every newly added _leaf_, we "canonicalize" the oldest leaf in the offchain db
+				// still using `(parent_hash, pos)` key.
+				let leaf_to_canon = leaf_index.saturating_sub(window_size);
+				let parent_hash_of_leaf = Self::parent_hash_of_leaf(leaf_to_canon, leaves);
+				let nodes_to_canon = Self::nodes_added_by_leaf(leaf_to_canon, leaves);
+				for pos in nodes_to_canon {
+					let key = Pallet::<T, I>::offchain_key(parent_hash_of_leaf, pos);
+					// FIXME: this approach also doesn't work because we can't read from offchain db
+					// in runtime context :((
+					let elem = vec![]; // this should have been: `offchain_index::get(&key)`
+					offchain_index::clear(&key);
+					let canon_key = Pallet::<T, I>::final_offchain_key(pos);
+					offchain_index::set(&canon_key, &elem);
```

Comment:
The idea was to keep a rolling window of `(parent_hash, pos)` leaf entries in the offchain db, with the window matching the one that provides `block_num -> block_hash` mappings in `frame_system`.

Once a leaf exits the window it would be "canonicalized" by switching its offchain db key from `(parent_hash, pos)` to simple `pos`.

This doesn't work however because there's no way to get leaf contents from offchain db while in runtime context.. so no way to `get(old_key) + clear(old_key) + set(new_key)` to effectively change the leaf's key in offchain db.
    
### Ideas:
1. move the "canonicalization" logic to offchain worker
2. enhance IndexingApi with `offchain::move(old_key, new_key)`
    This is weird, but correct, deterministic and safe AFAICT, so it could be exposed to runtime.

Reply:
Now have it working using offchain worker :+1: